src/App/Fossa/VSI/Fingerprint.hs

```diff
@@ -195,6 +170,30 @@ basicCStyleCommentStripC =
     .| filterC (not . Text.null)
     .| bufferedNewline Nothing
   where
+    -- For compatibility with the original version of this function we can't write a newline after the final line in the output, but we want newlines otherwise.
+    -- As we read through the input stream, instead of writing lines directly we'll buffer one at a time.
+    -- This way we can delay the decision of whether to write a trailing newline until we know if we're at the end of the input.
+    bufferedNewline buf = do
+      chunk <- await
+      case chunk of
+        -- Now that we're done reading the input stream, if there's a buffered output line write it *without a trailing newline*.
+        -- This is to maintain compatibility with the original version of this function.
+        -- We have to keep this compatible, because all of our fingerprint corpus relies on how this fingerprint function works.
+        Nothing -> case buf of
+          Nothing -> pure ()
+          Just bufferedLine -> do
+            yield bufferedLine
+            bufferedNewline Nothing
+
+        -- Here, we know we have a new line coming down the pipe.
+        -- If we had a previous line buffered, write it with a newline.
+        -- If not, store this new line in the buffer first.
+        Just line -> case buf of
+          Nothing -> bufferedNewline (Just line)
+          Just bufferedLine -> do
+            yield $ bufferedLine <> "\n"
+            bufferedNewline (Just line)
+
```

Comment:
[optional]: I think this could be made a little clearer by writing it as a `case (chunk, buf) of` expression.  As it currently is, this jumps around a lot and was hard to follow in terms of readability (programmatically, it's fine).

I'm thinking it could look a little more like this (note the ordering being consistent with actual execution)
```haskell
case (chunk, buf) of
  -- First line will land here, we always buffer the first line
  (Just line, Nothing) -> bufferedNewline (Just line)
  -- All lines between first and last will be yielded with newlines, and the NEXT line will be buffered
  (Just line, Just bufferedLine) -> do
    yield $ bufferedLine <> "\n"
    bufferedNewline (Just line)
  -- No next line, so the buffered line is the last one.
  (Nothing, Just bufferedLine) -> yield bufferedLine
  -- All lines have been written, now we can exit.
  -- Technically unreachable, since we don't recurse after yielding the last line.
  (Nothing, Nothing) -> pure ()
```

Reply:
But don't kill your comments about compat with old fingerprint data, those are obviously needed.

Reply:
Oooh, I like this. Good idea!